perf/src/recycler.rs

```diff
@@ -73,41 +181,132 @@ impl<T: Default + Reset + Sized> Recycler<T> {
+                    object_pool.total_allocated_count,
+                );
+
+                // If more than the shrink threshold of all allocated objects are sitting doing nothing,
+                // increment the `above_shrink_ratio_count`.
+                if object_pool.len() > object_pool.minimum_object_count as usize
+                    && object_pool.len() > shrink_threshold_count as usize
+                {
+                    object_pool.above_shrink_ratio_count += 1;
+                } else {
+                    object_pool.above_shrink_ratio_count = 0;
+                }
+
+                if object_pool.above_shrink_ratio_count as usize
+                    >= object_pool.max_above_shrink_ratio_count as usize
+                {
+                    let mut recycler_shrink_elapsed = Measure::start("recycler_shrink");
+                    // Do the shrink
+                    let target_size =
+                        std::cmp::max(object_pool.minimum_object_count, shrink_threshold_count);
+                    let ideal_num_to_remove = object_pool.total_allocated_count - target_size;
+                    for _ in 0..ideal_num_to_remove {
+                        if let Some(mut expired_object) = object_pool.object_pool.pop() {
+                            expired_object.unset_recycler();
+                            // Drop these outside of the lock because the Drop() implmentation for
+                            // certain objects like PinnedVec's can be expensive
+                            shrink_removed_objects.push(expired_object);
+                            // May not be able to shrink exactly `ideal_num_to_remove` objects sinc
+                            // in the case of new allocations, `total_allocated_count` is incremented
+                            // before the object is allocated (see `should_allocate_new` logic below).
+                            // This race allows a difference of up to the number of threads allocating
+                            // with this recycler.
+                            object_pool.total_allocated_count -= 1;
+                        } else {
+                            break;
+                        }
+                    }
+                    recycler_shrink_elapsed.stop();
+                    object_pool.above_shrink_ratio_count = 0;
+                    shrink_stats = Some(RecyclerShrinkStats {
+                        resulting_size: object_pool.total_allocated_count,
+                        target_size,
+                        ideal_num_to_remove,
+                        shrink_elapsed: recycler_shrink_elapsed.as_us(),
+                        // Filled in later
+                        drop_elapsed: 0,
+                    })
+                }
+            }
+
+            let reused_object = object_pool.object_pool.pop();
+            if reused_object.is_some() {
+                (reused_object, true, false, shrink_stats)
```

Comment:
nits: I don't think this is really simpler writing, arguably. However I can't resist to propose this: 

```rust
if let Some(reused_object) = object_pool.object_pool.pop() {
  (Some(reused_object), true, false, shrink_stats)
} ...
```

Reply:
fixed in separate `make_allocation_decision()` function~